swarm/src/handler.rs

```diff
@@ -338,94 +335,110 @@ impl<'a> ProtocolsChange<'a> {
+        existing_protocols: &mut HashSet<StreamProtocol>,
+        temp_protocols: &'a mut Vec<StreamProtocol>,
     ) -> Option<Self> {
-        let mut actually_removed_protocols = existing_protocols.intersection(to_remove).peekable();
-
-        actually_removed_protocols.peek()?;
+        temp_protocols.clear();
+        temp_protocols.extend(
+            to_remove
+                .into_iter()
+                .filter_map(|i| existing_protocols.take(&i)),
+        );
+
+        if temp_protocols.is_empty() {
+            return None;
+        }
 
-        Some(ProtocolsChange::Removed(ProtocolsRemoved {
-            protocols: Either::Right(actually_removed_protocols),
+        Some(Self::Removed(ProtocolsRemoved {
+            protocols: temp_protocols.iter(),
         }))
     }
 
     /// Compute the [`ProtocolsChange`]s required to go from `existing_protocols` to `new_protocols`.
-    pub(crate) fn from_full_sets(
-        existing_protocols: &'a HashSet<StreamProtocol>,
-        new_protocols: &'a HashSet<StreamProtocol>,
+    pub(crate) fn from_full_sets<T: AsRef<str>>(
+        existing_protocols: &mut HashMap<AsStrHashEq<T>, bool>,
+        new_protocols: impl IntoIterator<Item = T>,
+        temp_owner: &'a mut Vec<StreamProtocol>,
     ) -> SmallVec<[Self; 2]> {
-        if existing_protocols == new_protocols {
-            return SmallVec::new();
+        temp_owner.clear();
+        // to avoid looping trough the map to just set the booleans we use the fact that all of
+        // the booleans in the map have same value
+        let valid_value = !existing_protocols.values().next().copied().unwrap_or(false);
+        let mut count = Some(existing_protocols.len());
+        for new_protocol in new_protocols {
+            *existing_protocols
+                .entry(AsStrHashEq(new_protocol))
+                .or_insert_with_key(|k| {
+                    temp_owner.extend(StreamProtocol::try_from_owned(k.0.as_ref().to_owned()).ok());
+                    count = None;
+                    false
+                }) = valid_value;
+            count = count.map(|c| c - 1);
         }
 
-        let mut changes = SmallVec::new();
-
-        let mut added_protocols = new_protocols.difference(existing_protocols).peekable();
-        let mut removed_protocols = existing_protocols.difference(new_protocols).peekable();
+        if count == Some(0) {
+            return smallvec![];
+        }
 
-        if added_protocols.peek().is_some() {
-            changes.push(ProtocolsChange::Added(ProtocolsAdded {
-                protocols: added_protocols,
+        let added_count = temp_owner.len();
+        existing_protocols.retain(|k, v| {
+            if *v != valid_value {
+                temp_owner.push(StreamProtocol::try_from_owned(k.0.as_ref().to_owned()).unwrap());
+            }
+            *v == valid_value
+        });
+
+        let (added, removed) = temp_owner.split_at(added_count);
```

Comment:
I don't yet understand how your performance optimisation works but I am wondering if it would get simpler if you'd have two separate buffers? You could keep one `Vec` in the `Connection` and split it into two `&mut [StreamProtocol]` using `split_at` before you call `from_full_sets`.

Reply:
You cannot push to slice and I cannot predict how many protocols were removed and added before calling this function. I use one Vec because it reduces amount of things that need to exist and be passed around.

Reply:
Thanks for the explanation!